src/init.c

```diff
@@ -378,6 +383,51 @@ shmem_internal_heap_postinit(void)
               shmem_internal_data_base, shmem_internal_data_length);
 
 #ifdef HAVE_SCHED_GETAFFINITY
+#ifdef USE_HWLOC
+    ret = hwloc_topology_init(&shmem_internal_topology);
+    if (ret < 0) {
+        RETURN_ERROR_MSG("hwloc_topology_init failed (%s)\n", strerror(errno));
+    }
+
+    ret = hwloc_topology_set_io_types_filter(shmem_internal_topology, HWLOC_TYPE_FILTER_KEEP_ALL);
+	if (ret < 0) {
+        RETURN_ERROR_MSG("hwloc_topology_set_io_types_filter failed (%s)\n", strerror(errno));
+    }
+
+    ret = hwloc_topology_load(shmem_internal_topology);
+    if (ret < 0) {
+        RETURN_ERROR_MSG("hwloc_topology_load failed (%s)\n", strerror(errno));
+    }
+#if defined HWLOC_ENFORCE_SINGLE_SOCKET || defined HWLOC_ENFORCE_SINGLE_NUMA_NODE
+    hwloc_bitmap_t bindset = hwloc_bitmap_alloc();
+    hwloc_bitmap_t bindset_all = hwloc_bitmap_alloc();
+    hwloc_bitmap_t bindset_covering_obj = hwloc_bitmap_alloc();
+
+    ret = hwloc_get_proc_last_cpu_location(shmem_internal_topology, getpid(), bindset, HWLOC_CPUBIND_PROCESS);
+    if (ret < 0) {
+        RETURN_ERROR_MSG("hwloc_get_proc_last_cpu_location failed (%s)\n", strerror(errno));
+    }
+
+    ret = hwloc_get_proc_cpubind(shmem_internal_topology, getpid(), bindset_all, HWLOC_CPUBIND_PROCESS);
+    if (ret < 0) {
+        RETURN_ERROR_MSG("hwloc_get_proc_cpubind failed (%s)\n", strerror(errno));
+    }
+#ifdef HWLOC_ENFORCE_SINGLE_SOCKET
+    hwloc_obj_t covering_obj = hwloc_get_next_obj_covering_cpuset_by_type(shmem_internal_topology, bindset, HWLOC_OBJ_PACKAGE, NULL);
+    if (!covering_obj) RETURN_ERROR_MSG("hwloc_get_next_obj_covering_cpuset_by_type failed (could not detect object of type 'HWLOC_OBJ_PACKAGE' in provided cpuset)\n");
+#else //HWLOC_ENFORCE_SINGLE_NUMA_NODE
+    hwloc_obj_t covering_obj = hwloc_get_next_obj_covering_cpuset_by_type(shmem_internal_topology, bindset, HWLOC_OBJ_NUMANODE, NULL);
+    if (!covering_obj) RETURN_ERROR_MSG("hwloc_get_next_obj_covering_cpuset_by_type failed (could not detect object of type 'HWLOC_OBJ_NUMANODE' in provided cpuset)\n");
+#endif
+    hwloc_bitmap_and(bindset_covering_obj, bindset_all, covering_obj->cpuset);
+    hwloc_set_proc_cpubind(shmem_internal_topology, getpid(), bindset_covering_obj, HWLOC_CPUBIND_PROCESS); //Include HWLOC_CPUBIND_STRICT in flags?
```

Comment:
Check return value?

Reply:
Good question about `HWLOC_CPUBIND_STRICT`.  I'm ok leaving it out with the comment for now, though my guess is that for most HPC use-cases the flag would be appropriate (i.e., do the assignment regardless if the CPU(s) is busy at init time).

Reply:
@philipmarshall21 - looks like this PR is all set except maybe this item.  Should we check the return value of `hwloc_set_proc_cpubind`?  It looks like the docs are a little rough here overall, but this suggests it's 0 or -1 like many of the other routines:
https://hwloc.readthedocs.io/en/stable/group__hwlocality__cpubinding.html